creusot-contracts/src/logic/ord.rs

```diff
@@ -0,0 +1,83 @@
+use crate as creusot_contracts;
+use creusot_contracts_proc::*;
+
+use crate::logic::EqLogic;
+pub use std::cmp::Ordering;
+
+pub trait OrdLogic: EqLogic {
+    #[logic]
+    fn cmp_log(self, _: Self) -> Ordering;
+
+    #[logic]
+    fn le_log(self, o: Self) -> bool {
+        match self.cmp_log(o) {
+            Ordering::Greater => false,
+            _ => true,
+        }
+    }
+
+    #[logic]
+    #[ensures(x.le_log(y) === ! (x.cmp_log(y) === Ordering::Greater))]
+    fn cmp_le_log(x: Self, y: Self);
+
+    #[logic]
+    fn lt_log(self, o: Self) -> bool {
+        match self.cmp_log(o) {
+            Ordering::Less => true,
+            _ => false,
+        }
+    }
+
+    #[logic]
+    #[ensures(x.le_log(y) === (x.cmp_log(y) === Ordering::Less))]
+    fn cmp_lt_log(x: Self, y: Self);
+
+    #[logic]
+    fn ge_log(self, o: Self) -> bool {
+        match self.cmp_log(o) {
+            Ordering::Less => false,
+            _ => true,
+        }
+    }
+
+    #[logic]
+    #[ensures(x.le_log(y) === ! (x.cmp_log(y) === Ordering::Less))]
+    fn cmp_ge_log(x: Self, y: Self);
+
+    #[logic]
+    fn gt_log(self, o: Self) -> bool {
+        match self.cmp_log(o) {
+            Ordering::Greater => true,
+            _ => false,
+        }
+    }
+
+    #[logic]
+    #[ensures(x.le_log(y) === (x.cmp_log(y) === Ordering::Greater))]
+    fn cmp_gt_log(x: Self, y: Self);
+
+    #[logic]
+    #[ensures(x.cmp_log(x) === Ordering::Equal)]
+    fn refl(x: Self);
+
+    #[logic]
+    #[requires(x.cmp_log(y) === o)]
+    #[requires(y.cmp_log(z) === o)]
+    #[ensures(x.cmp_log(z) === o)]
+    fn trans(x: Self, y: Self, z: Self, o: Ordering);
+
+    #[logic]
+    #[requires(x.cmp_log(y) === Ordering::Less)]
+    #[ensures(y.cmp_log(x) === Ordering::Greater)]
+    fn antisym1(x: Self, y: Self);
+
+    #[logic]
+    #[requires(x.cmp_log(y) === Ordering::Greater)]
+    #[ensures(y.cmp_log(x) === Ordering::Less)]
+    fn antisym2(x: Self, y: Self);
+
+    #[logic]
+    #[requires(x.cmp_log(y) === Ordering::Equal)]
+    #[ensures(y.cmp_log(x) === Ordering::Equal)]
+    fn symmetry(x: Self, y: Self);
```

Comment:
Instead of that, we should perhaps have a lemma relating `cmp_log` with `eq_log`, and rely on symmetry of equality instead.